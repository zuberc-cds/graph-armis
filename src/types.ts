// Providers often supply types with their API libraries.

export interface AcmeUser {
  id: string;
  name: string;
}

export interface AcmeGroup {
  id: string;
  name: string;
  users?: Pick<AcmeUser, 'id'>[];
}

export interface ArmisDevice {
  id: number;
  name: string;
  category: string;
  type: string;
  model: string;
  manufacturer: string;
  operatingSystem: string;
  operatingSystemVersion: string;
  lastSeen: string;
}

export interface ArmisSite {
  id: number;
  lat: number;
  lng: number;
  location: string;
  name: string;
  parentId: string;
}

export interface ArmisAlert extends ArmisFinding {
  alertId: number;
  title: string;
  classification: string;
  type: string;
  policyId: number;
  policyTitle: string;
  time: string;
  deviceIds: number[];
}

export interface ArmisVulnerability extends ArmisFinding {
  cveUid: string;
  affectedDevicesCount: number;
  attackComplexity: string;
  attackVector: string;
  availabilityImpact: string;
  cvssScore: number;
  epssPercentile: number;
  epssScore: number;
  exploitabilityScore: number;
  impactScore: number;
  integrityImpact: string;
  publishedDate: string;
  score: number;
  commonName: string;
  threatTags: string[];
}

export interface ArmisFinding {
  id: string;
  description: string;
  severity: string;
  status: string;
}

export type ArmisDeviceVulnerability = {
  cveUid: string;
};

// Those can be useful to a degree, but often they're just full of optional
// values. Understanding the response data may be more reliably accomplished by
// reviewing the API response recordings produced by testing the wrapper client
// (./client.ts). However, when there are no types provided, it is necessary to define
// opaque types for each resource, to communicate the records that are expected
// to come from an endpoint and are provided to iterating functions.

/*
import { Opaque } from 'type-fest';
export type AcmeUser = Opaque<any, 'AcmeUser'>;
export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
*/
