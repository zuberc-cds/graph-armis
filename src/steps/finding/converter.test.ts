import { ArmisAlert, ArmisVulnerability } from '../../types';
import {
  createAlertFindingEntity,
  createVulnerabilityFindingEntity,
} from './converter';

describe('#createAlertFindingEntity', () => {
  test('should convert to alert entity', () => {
    const alertEntity: ArmisAlert = {
      alertId: 219,
      status: 'Unhandled',
      description:
        'When data is sent encrypted on the network, it dramatically reduces the impact of that data being intercepted by someone that should not have it. Not all data needs to be encrypted, especially when that data is traversing an internal network. Unencrypted protocols can be potentially intercepted resulting in sensitive company information being stolen.',
      classification: 'Security - Other',
      deviceIds: [133],
      policyId: 110807,
      policyTitle: '[Risk] Unencrypted Traffic in Use',
      severity: 'Medium',
      time: '2023-11-26T19:06:58.601691+00:00',
      title: '[Risk] Unencrypted Traffic in Use',
      type: 'System Policy Violation',
    };
    expect(createAlertFindingEntity(alertEntity)).toMatchSnapshot();
  });

  test('should convert to vulnerability entity', () => {
    const vulnerabilityEntity: ArmisVulnerability = {
      affectedDevicesCount: 3,
      attackComplexity: 'Low',
      attackVector: 'Network',
      availabilityImpact: 'High',
      commonName: null,
      cveUid: 'CVE-2022-32845',
      cvssScore: 10,
      description:
        'This issue was addressed with improved checks. This issue is fixed in watchOS 8.7, iOS 15.6 and iPadOS 15.6, macOS Monterey 12.5. An app may be able to break out of its sandbox.',
      epssPercentile: 0.53,
      epssScore: 0.00166,
      exploitabilityScore: 3.9,
      id: 'CVE-2022-32845',
      impactScore: 6,
      integrityImpact: 'High',
      publishedDate: '2022-09-23T19:15:00+00:00',
      score: 10,
      severity: 'Critical',
      status: 'Open',
      threatTags: null,
    };
    expect(
      createVulnerabilityFindingEntity(vulnerabilityEntity),
    ).toMatchSnapshot();
  });
});
