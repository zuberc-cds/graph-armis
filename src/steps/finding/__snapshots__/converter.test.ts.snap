// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createAlertFindingEntity should convert to alert entity 1`] = `
{
  "_class": [
    "Finding",
  ],
  "_key": "armis-alert-219",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "alertId": 219,
        "classification": "Security - Other",
        "description": "When data is sent encrypted on the network, it dramatically reduces the impact of that data being intercepted by someone that should not have it. Not all data needs to be encrypted, especially when that data is traversing an internal network. Unencrypted protocols can be potentially intercepted resulting in sensitive company information being stolen.",
        "deviceIds": [
          133,
        ],
        "policyId": 110807,
        "policyTitle": "[Risk] Unencrypted Traffic in Use",
        "severity": "Medium",
        "status": "Unhandled",
        "time": "2023-11-26T19:06:58.601691+00:00",
        "title": "[Risk] Unencrypted Traffic in Use",
        "type": "System Policy Violation",
      },
    },
  ],
  "_type": "armis_finding_alert",
  "active": undefined,
  "category": "System Policy Violation",
  "classification": "Security - Other",
  "createdOn": undefined,
  "description": "When data is sent encrypted on the network, it dramatically reduces the impact of that data being intercepted by someone that should not have it. Not all data needs to be encrypted, especially when that data is traversing an internal network. Unencrypted protocols can be potentially intercepted resulting in sensitive company information being stolen.",
  "displayName": "[Risk] Unencrypted Traffic in Use",
  "name": "[Risk] Unencrypted Traffic in Use",
  "numericSeverity": 1,
  "open": false,
  "policyId": 110807,
  "policyTitle": "[Risk] Unencrypted Traffic in Use",
  "reportedOn": 1701025618601,
  "severity": "Medium",
  "status": "Unhandled",
}
`;

exports[`#createAlertFindingEntity should convert to vulnerability entity 1`] = `
{
  "_class": [
    "Vulnerability",
  ],
  "_key": "armis-vulnerability-CVE-2022-32845",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "affectedDevicesCount": 3,
        "attackComplexity": "Low",
        "attackVector": "Network",
        "availabilityImpact": "High",
        "commonName": null,
        "cveUid": "CVE-2022-32845",
        "cvssScore": 10,
        "description": "This issue was addressed with improved checks. This issue is fixed in watchOS 8.7, iOS 15.6 and iPadOS 15.6, macOS Monterey 12.5. An app may be able to break out of its sandbox.",
        "epssPercentile": 0.53,
        "epssScore": 0.00166,
        "exploitabilityScore": 3.9,
        "id": "CVE-2022-32845",
        "impactScore": 6,
        "integrityImpact": "High",
        "publishedDate": "2022-09-23T19:15:00+00:00",
        "score": 10,
        "severity": "Critical",
        "status": "Open",
        "threatTags": null,
      },
    },
  ],
  "_type": "armis_vulnerability",
  "active": undefined,
  "blocking": false,
  "category": "Network",
  "createdOn": undefined,
  "description": "This issue was addressed with improved checks. This issue is fixed in watchOS 8.7, iOS 15.6 and iPadOS 15.6, macOS Monterey 12.5. An app may be able to break out of its sandbox.",
  "displayName": "armis-vulnerability-CVE-2022-32845",
  "exploitability": 3.9,
  "impact": 6,
  "impacts": [
    "devices",
  ],
  "name": "armis-vulnerability-CVE-2022-32845",
  "numericSeverity": 3,
  "open": true,
  "production": true,
  "public": false,
  "reportedOn": 1663960500000,
  "score": 10,
  "severity": "Critical",
  "status": "Open",
  "vector": "Network",
}
`;
